### PR TITLE
[QEff. Finetune] Fixed reporting of single value of loss and ppl across devices.

### PR DESCRIPTION
**Pull request by quic-meetkuma**
- Fixed reporting of single value of loss and ppl across devices.
- Minor refactoring changes and variable name changes to make it consistent.

### CODE REVIEW

File: QEfficient/finetune/utils/train_utils.py [hunk 1]
```diff
@@ -353,35 +354,42 @@ def train(
...
+            )
+            # Print evaluation metrics
+            print(
+                f"Epoch {epoch + 1}: Eval Loss: {eval_loss.detach().cpu():.4f}, Eval metric: {eval_metric.detach().cpu():.4f}"
+            )
+            if eval_loss < best_val_loss:
+                best_val_loss = eval_loss
+                print(f"best eval loss on epoch {epoch + 1} is {best_val_loss:.4f}")
 
+            if local_rank == 0:
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:386]: Good catch, will update!

File: QEfficient/finetune/utils/train_utils.py [hunk 2]
```diff
@@ -512,18 +507,19 @@ def evaluation_helper(model, train_config, eval_dataloader, device):
 
             eval_loss += loss.detach().float()
     # Compute average loss and metric
-    eval_epoch_loss = (
-        0.0 if eval_loss == 0.0 else eval_loss / (step + 1 - num_dummy_samples / train_config.val_batch_size)
-    )
+    eval_loss = 0.0 if eval_loss == 0.0 else eval_loss / (step + 1 - num_dummy_samples / train_config.val_batch_size)
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:510]: Check the other variables being returned from this function. Made the names consistent.

**Review by quic-mamta — APPROVED** [review]

File: QEfficient/finetune/utils/train_utils.py [hunk 3]
```diff
@@ -421,17 +418,12 @@ def train(
...
-    results["avg_train_metric"] = avg_train_metric
-    results["avg_train_loss"] = avg_train_loss
+    results["last_epoch_train_loss"] = train_epoch_loss
+    results["last_epoch_train_metric"] = train_epoch_metric
     if train_config.run_validation:
-        results["avg_eval_metric"] = avg_eval_metric
-        results["avg_eval_loss"] = avg_eval_loss
+        results["last_epoch_eval_loss"] = eval_loss
+        results["last_epoch_eval_metric"] = eval_metric
     results["avg_epoch_time"] = avg_epoch_time
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:427]: tests also needs to be updated for this variable name changes in results dict.

File: QEfficient/finetune/utils/train_utils.py [hunk 4]
```diff
@@ -130,7 +130,6 @@ def train(
                 continue
 
         logger.log_rank_zero(f"Starting epoch {epoch + 1}/{train_config.num_epochs}")
-        logger.log_rank_zero(f"train_config.max_train_step: {train_config.max_train_step}")
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:133]: we should keep this under if train_config.max_train_step > 0 :

File: QEfficient/finetune/utils/train_utils.py [hunk 5]
```diff
@@ -284,54 +283,63 @@ def train(
         epoch_end_time = time.perf_counter() - epoch_start_time
         epoch_times.append(epoch_end_time)
 
-        if loss_0_counter.item() == train_config.convergence_counter:
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/train_utils.py:287]: @quic-swatia , is this line required here?

File: QEfficient/finetune/utils/dataset_utils.py [hunk 6]
```diff
@@ -77,24 +84,26 @@ def padding_dataset(train_config, dataset, batch_size):
 
     dummy_row = next(iter(dataset))
     dummy_row["labels"] = torch.tensor([-100] * len(dummy_row["labels"]))
-    padding_size = 0
```
Comment (quic-mamta) [inline QEfficient/finetune/utils/dataset_utils.py:80]: this is required as discussed @quic-swatia

File: QEfficient/finetune/utils/train_utils.py [hunk 7]
```diff
@@ -284,54 +283,63 @@ def train(
...
+            train_epoch_loss /= get_num_ddp_devices()
+            dist.all_reduce(train_epoch_metric, op=dist.ReduceOp.SUM)
+            train_epoch_metric /= get_num_ddp_devices()
+
         # Update the learning rate as needed
         lr_scheduler.step()
 
         if train_config.run_validation:
-            eval_epoch_loss, eval_metric, temp_val_loss, temp_step_metric = evaluation_helper(
+            eval_epoch_loss, eval_epoch_metric, step_loss, step_metric = evaluation_helper(
```
Comment (quic-swatia) [inline QEfficient/finetune/utils/train_utils.py:318]: step_loss should be renamed to eval_step_loss to match train_step_loss. Similarly, step_metric should be renamed.

File: QEfficient/finetune/utils/train_utils.py [hunk 8]
```diff
@@ -205,23 +204,23 @@ def train(
...
             if train_config.save_metrics:
                 train_step_loss.append(loss.detach().float().item())
                 if train_config.task_type == "seq_classification":
-                    step_metric_val = float(acc_helper.compute())
+                    step_metric_value = float(acc_helper.compute())
                 else:
-                    step_metric_val = float(torch.exp(loss.detach().float()))
-                train_step_metric.append(step_metric_val)
+                    step_metric_value = float(torch.exp(loss.detach().float()))
+                train_step_metric.append(step_metric_value)
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py:223]: We can but, if acc_helper.compute() or loss.detach() or torch.exp() throws an error the logs would say that train_step_metric.append -> line has the error. In that case it will be difficult to narrow down the error.

File: QEfficient/finetune/utils/dataset_utils.py (hunk 6 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/dataset_utils.py:80]: This is handled by if remainder == 0 return dataset

File: QEfficient/finetune/utils/train_utils.py (hunk 7 above)
Comment (quic-mamta) [reply QEfficient/finetune/utils/train_utils.py:323]: can we use epoch_loss, epoch_metric, step_loss, step_metric here and while saving metrics at line 334 we use, eval_epoch_loss, eval_epoch_metric, eval_step_loss, eval_step_metric,

File: QEfficient/finetune/utils/train_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply QEfficient/finetune/utils/train_utils.py]: Yes, updated all the instances and make it more consistent throughout the code. Thanks for the suggestion.

File: tests/finetune/test_finetune.py [hunk 9]
```diff
@@ -18,6 +18,7 @@
 import QEfficient
 import QEfficient.cloud.finetune
 from QEfficient.cloud.finetune import main as finetune
+from QEfficient.finetune.utils.helper import Device, Task_Mode
 
 alpaca_json_path = Path.cwd() / "alpaca_data.json"
```
Comment (quic-mamta) [inline tests/finetune/test_finetune.py:23]: can we use os.getcwd() so new module will not be imported.

File: tests/finetune/test_finetune.py [hunk 10]
```diff
@@ -85,11 +86,11 @@ def download_alpaca():
...
-        "qaic",  # device
-        0.63060283,  # expected_train_loss
-        0.55554199,  # expected_train_metric
-        0.61503016,  # expected_eval_loss
-        0.70825195,  # expected_eval_metric
+        Device.QAIC,  # device
+        0.6566,  # expected_train_loss
+        0.5055,  # expected_train_metric
+        0.6188,  # expected_eval_loss
+        0.6904,  # expected_eval_metric
```
Comment (quic-mamta) [inline tests/finetune/test_finetune.py:93]: can we change intermediate_step_save at line 85 and other places to 10, we don't need to save the weights for testing after each step.

File: tests/finetune/test_finetune.py [hunk 11]
```diff
@@ -151,15 +152,19 @@ def test_finetune_llama(
...
+    assert np.allclose(results["last_epoch_eval_loss"], expected_eval_loss, atol=1e-3), "Eval loss is not matching."
+    assert np.allclose(results["last_epoch_eval_metric"], expected_eval_metric, atol=1e-3), (
+        "Eval metric is not matching."
+    )
     assert results["avg_epoch_time"] < 60, "Training should complete within 60 seconds."
 
     train_config_spy.assert_called_once()
     generate_dataset_config_spy.assert_called_once()
-    if task_type == "generation":
+    if task_mode == "generation":
```
[the remark this answers is not in the document]
Comment (quic-meetkuma) [reply tests/finetune/test_finetune.py:167]: Thanks for spotting, missed when updating other instances.
